### PR TITLE
pubtools-exodus-push: Use pushsource library to load content [RHELDST-10120]

### PR DESCRIPTION
**Pull request by crungehottman**
The pubtools-exodus-push entry point now uses the pushsource
library to load content. This means that content pushed using
pubtools-exodus-push must be staged in a pushsource-compatible
staging structure.

The entry point can be invoked like so:

`$ pubtools-exodus-push $PUSHSOURCE_BACKEND:$STAGING_ROOT`

For example:

```
$ tree /home/crungeho/example-staging-dir
/home/crungeho/example-staging-dir
├── kickstart-repo-s390x
│   └── RAW
│       └── test.txt
└── kickstart-repo-x86_64
    └── RAW
        └── test.txt
```

`$ pubtools-exodus-push staged:/home/crungeho/example-staging-dir`

### CODE REVIEW

File: pubtools/exodus/_tasks/push.py [hunk 1]
```diff
@@ -14,12 +16,12 @@ def add_args(self):
...
 
         self.parser.add_argument(
-            "sync",
-            nargs="+",
+            "source",
             help=(
-                """A comma-separated pair of paths to sync: SRC,DEST.\n"""
-                """Where SRC is a local path to a file or directory and DEST\n"""
-                """is the remote destination for sync."""
+                """$PUSHSOURCE_BACKEND:$STAGING_ROOT, where $PUSHSOURCE_BACKEND is a valid """
```
Comment (rohanpm) [inline pubtools/exodus/_tasks/push.py]: This help text isn't right, since by loading the content through pushsource we support *all* pushsource backends and not just those which use a $STAGING_ROOT.  Could it be more general please, like just "Source(s) of content to be pushed" ?  The example could be kept as a hint as to which sources are useful in practice.

Note I'm also maintaining internal doc https://docs.engineering.redhat.com/x/60QiDQ , that's a place which can be used to add more specific detail.  Maybe you want to document the new custom backend there.

File: pubtools/exodus/_tasks/push.py [hunk 2]
```diff
@@ -30,13 +32,21 @@ def run(self):
         publish_id = str(publish.get("id"))
         LOG.info("Publish ID: %s", publish_id)
 
-        for item in self.args.sync:
+        push_items = []
+        with Source.get(self.args.source) as source:
+            for push_item in source:
```
Comment (rohanpm) [inline pubtools/exodus/_tasks/push.py]: Could you please make this tolerate (warn or ignore) unknown push item types? It's suggested in https://release-engineering.github.io/pushsource/userguide.html#processing-push-items to ensure the code doesn't break if pushsource library later adds support for new kinds of items.

Based on the way the code works later on, maybe rather than doing it by an isinstance check, the right way would be something like this:

```python
if push_item.src and len(push_item.dest) == 1:
    push_items.append(push_item)
else:
    LOG.warning("Skipping unsupported item: %s", push_item)
```

Basically if the item contains a src (file or directory) and a single dest, then we know how to handle it.  If not, then we don't.

File: pubtools/exodus/_tasks/push.py [hunk 3]
```diff
@@ -30,13 +32,21 @@ def run(self):
         publish_id = str(publish.get("id"))
         LOG.info("Publish ID: %s", publish_id)
 
-        for item in self.args.sync:
+        push_items = []
+        with Source.get(self.args.source) as source:
```
Comment (rohanpm) [inline pubtools/exodus/_tasks/push.py]: While I'm not sure this will matter much in practice, maybe not (depends on the size of the pushes), it's generally intended that you should be pushing the items as you find them... not waiting until you find all items.

For instance some larger staging directories themselves can take 30+ minutes just to iterate over, especially if the connection between you and the NFS share isn't super fast.  And there is no reason to wait all that time before starting the first exodus-rsync command, when you could immediately start it as soon as you've got the first item.

File: pubtools/exodus/_tasks/push.py [hunk 4]
```diff
@@ -30,32 +29,40 @@ def run(self):
...
-            )
-            for line in proc.stdout:
-                LOG.info(line.strip())
-
-            ret = proc.wait()
-            if ret != 0:
-                raise RuntimeError("Exodus push failed")
+        with Source.get(self.args.source) as source:
+            for item in source:
+                if item.src and len(item.dest) == 1:
```
Comment (rohanpm) [inline pubtools/exodus/_tasks/push.py]: Minor: if you feel like this is getting a bit too nested (which I personally do), you can always put the iteration logic in its own method something like this:

```python
@property
def push_items(self):
    with Source.get(self.args.source) as source:
        for item in source:
            if item.src and len(item.dest) == 1:
                yield item
            else:
                LOG.warning(...)
```

The main loop here would then look cleaner like:

```python
for item in self.push_items:
    LOG.debug("Processing %s", item)
    # ...the stuff for running exodus-rsync
```

**Review by rohanpm — DISMISSED** [review]

File: tests/test_exodus_push_task.py [hunk 5]
```diff
@@ -29,12 +32,12 @@ def test_exodus_push_typical(mock_popen, successful_gw_task, caplog):
...
     mock_popen.return_value.wait.return_value = 0
 
+    src = os.path.join(TEST_DATA, "source-1")
     args = [
         "--debug",
         "-vvv",
         "--dry-run",
-        "/path/to/my-source,/path/to/my-dest",
-        "/path/to/other-source,/path/to/other-dest",
+        "staged:%s" % src,
```
Comment (negillett) [inline tests/test_exodus_push_task.py:41]: Will content always be in 'staged'? I don't see any other examples.

Comment (negillett) [reply tests/test_exodus_push_task.py:41]: ~~I'm assuming that format is for pushsource~~ (I reread the commit body). ~~If the endpoint isn't doing anything with the backend name and it's always "staged", is the arg just enforcing that format for compatibility with other, similar tools?~~

Comment (negillett) [reply tests/test_exodus_push_task.py:41]: With my confusion about pushsource aside, the question I'm getting to is: if there are other backends, should we test some besides "staged" here as well or would that just be duplicating tests from pushsource?

Comment (rohanpm) [reply tests/test_exodus_push_task.py:41]: The code should aim to work with any pushsource backend, even though we only expect to use it with staged & cdn-utils-staged for the initial use-cases.  With the way libraries are divided up now, the idea is that code for pushing content should be totally independent from the code for looking up the content to be pushed, such that you can mix and match them in whatever way is needed.  It's reasonable for tests to generally use 'staged' source though because that's an easy way to set up realistic inputs, given that you can just create a certain directory structure and don't need to involve any remote systems.

A related thought I had while looking at test_exodus_push_unexpected_push_item, though ultimately decided it wasn't important enough to mention.  That test doesn't need to be mocking Source.get, since pushsource library is designed in a way which allows you to register your own backend.  You can even do it in a single line returning a list of arbitrary push items, like:

```python
    Source.register_backend("test", lambda: [PushItem(name="quux")])
```

...and you could then invoke the entry point with 'test:' as the source and see how it behaves.  So you really don't need to mock.

IMO changing it not to mock isn't strictly needed, but I mention this since if there was at least one test working like that, it would make it clearer that the code should work with arbitrary pushsource backends and not be specific to 'staged'.

**Review by rohanpm — DISMISSED** [review]

**Review by rohanpm — APPROVED** [review]
